src/jit/jitstd/algorithm.h

```diff
@@ -46,4 +46,191 @@ Function for_each(InputIterator first, InputIterator last, Function func)
+        // Note that first and last have alreay been partitioned so the loops below
+        // start by moving the iterator to the next position of interest.
+        RandomAccessIterator newFirst = first;
+        RandomAccessIterator newLast = last;
+
+        for (;;)
+        {
+            // Find newFirst such that *newFirst >= *pivot.
+            //
+            // less(*pivot, *pivot) is expected to return false so we should stop
+            // if we reach the pivot. However, the current JIT uses of std::sort
+            // have relatively expensive "less" predicates while the iterators are
+            // just pointers so they are cheap to compare. Thus it's best to check
+            // for the newFirst == pivot case and skip a "less" call.
+            //
+            // It's not possible for newFirst to go past the end of the sort range:
+            //   - If newFirst reaches the pivot before newLast then the pivot is
+            //	   swapped to the right and we'll stop again when we reach it.
+            //   - If newLast reaches the pivot before newFirst then the pivot is
+            //	   swapped to the left and the value at newFirst will take its place
+            //     to the right so less(newFirst, pivot) will again be false when the
+            //     old pivot's position is reached.
+            do
+            {
+                ++newFirst;
+            } while ((newFirst != pivot) && less(*newFirst, *pivot));
+
+            // Find newLast such that *newLast <= *pivot.
+            //
+            // Like above, this stops when the pivot is reached and also does not
+            // go before the start of the sort range.
+            do
+            {
+                --newLast;
+            } while ((newLast != pivot) && less(*pivot, *newLast));
+
+            // If newFirst reaches newLast then we're done.
+            if (newFirst >= newLast)
+            {
+                break;
+            }
+
+            // We now have *newLast <= *pivot <= *newFirst so we need to swap
+            // *newFirst and *newLast.
+            swap(*newFirst, *newLast);
+
+            // pivot is an iterator and not the actual value, if the value gets
+            // swapped we need to update the iterator to point to the new place.
+            if (pivot == newFirst)
+            {
+                pivot = newLast;
+            }
+            else if (pivot == newLast)
+            {
+                pivot = newFirst;
+            }
+        }
+
+        RandomAccessIterator leftFirst = first;
+        RandomAccessIterator leftLast = newLast;
+        RandomAccessIterator rightFirst = newLast + 1;
+        RandomAccessIterator rightLast = last;
+
+        assert(depth < _countof(firstStack));
+
+        // Ideally, the 2 partitions should have the same size, that would guarantee
+        // log2(n) stack space. If that's not the case then push the larger partition
```

Comment:
What is called `tail recursion` here?

Reply:
Hmm, presumably I should have said "**manual** tail recursion" or something similar since obviously there's no call/recursion in here, it has already been turned it into a loop.